File: Array/java/kthLargest.java
package Array;

import java.io.util.*;

class Solution {
    // O(N) best case / O(N^2) worst case running time + O(1) memory

    public int findKthLargest(int[] nums, int k) {

        return quickSelect(nums, 0, nums.length - 1, nums.length - k);

    }

    private int quickSelect(int[] nums, int s, int e, int k) {
        int random = new Random().nextInt((e - s) + 1) + s;
        swap(nums, random, e);

        int j = s, pivot = nums[e];
        for (int i = s; i < e; i++) {
            if (nums[i] <= pivot) {
                swap(nums, j, i);
                j++;
            }
        }
        swap(nums, j, e);

        if (j == k) {
            return nums[j];
        }
        // must be on the left part of pivot
        if (j > k) {
            return quickSelect(nums, s, j - 1, k);
        }
        // must be to the right of the pivot
        return quickSelect(nums, j + 1, e, k);

    }

    private void swap(int[] nums, int a, int b) {
        int temp = nums[a];
        nums[a] = nums[b];
        nums[b] = temp;
    }
}

public class kthLargest {
    // O(N) best case / O(N^2) worst case running time + O(1) memory

    public int findKthLargest(int[] nums, int k) {
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        for (int num : nums) {
            pq.offer(num);

            System.out.println(pq);
            if (pq.size() > k) {
                pq.poll();
            }
        }

        return pq.peek();
    }

}

File: Array/kthLargest.js

/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
 var findKthLargestSimple = function(nums, k) {
    let arr = nums.sort((a, b) => a - b); // [1, 2, 3, 4, 5, 6]
    return arr[nums.length - k] // arr[6 - 2] == 5
};



/* O(N) */
var findKthLargest = function(nums, k) {

    return quickSelect(nums, 0,nums.length-1, k);
    
    function quickSelect(nums,start,end,k){
        //choose a randomPivot and swap it to the end of the array
        let randomPivot = Math.floor(Math.random()*(end-start+1))+start;
        swap(nums,randomPivot, end);
        
        let pivotIndex = partition(nums, start, end);

        //if k is smaller than the target is in the right side
        if(k < nums.length - pivotIndex){
            return quickSelect(nums, pivotIndex+1, end, k);
        }
        
        //if k is smaller than the target is in the right side
        else if(k > nums.length - pivotIndex){
            return quickSelect(nums, start, pivotIndex-1, k);
        }
        
        //if k === nums.length - pivotIndex
        return nums[pivotIndex];
    }
    
    function partition(nums,start,end){
       const pivot = nums[end];
        let i = start;
        let j = end - 1;
        
        // move all numbers smaller, left of pivot, and larger to the right of pivot
        while(i<=j) {
            while(nums[i] < pivot)
                i++;
            while(nums[j]>pivot)
                j--;
            
            if(i<=j){
                swap(nums, i, j);
                i++;
                j--;
            }
        }
        // swap pivot to its final position
        swap(nums, i, end);
        return i;
    }
    
   function swap(nums,i,j){
        const temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
        // [nums[i], nums[j]] = [nums[j], nums[i]];

    }
};

